### PR TITLE
New bs4_book() template

### PR DESCRIPTION
**Pull request by hadley**
To do before merging: 

* [x] Remove customisations to bookdown book that I used for testing
* [x] Uncomment `common_format_config()`
* [x] Add news bullet

### CODE REVIEW

File: R/bs4_book.R [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+                     extra_dependencies = NULL
+                     ) {
+  check_packages(c("bootstraplib", "downlit", "jsonlite", "xml2"))
+
+  # Allow theme specification in yaml metadata
+  if (!inherits(theme, "bs_theme")) {
+    theme <- do.call(bs4_book_theme, theme)
+  }
+
+  config <- rmarkdown::html_document(
```
Comment (atusy) [inline R/bs4_book.R:53]:
```suggestion
  config <- html_document2(
```

I guess html_document2 is the better choice as we want cross-referencing feature.

Comment (hadley) [reply R/bs4_book.R:53]: I don't think that's correct — I'm manually restoring the cross-references using the same technique as `gitbook()`.

Comment (atusy) [reply R/bs4_book.R:53]: oh sorry, and i see.

File: R/bs4_book.R [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+#' * R syntax highlighting and autolinking by
+#'   [downlit](http://downlit.r-lib.org/) is paired with a accessible
+#'   theme designed by Alison Hill.
+#'
+#' * The ability to customise colours and fonts through
+#'   [bootstraplib](https://rstudio.github.io/bootstraplib)
+#'
+#' @section Limitations:
+#'
+#' * `bs_book4()` is designed specifically for books that use one chapter per
```
Comment (cderv) [inline R/bs4_book.R:29]: Just seen this typo while looking at the on-going PR
```suggestion
#' * `bs4_book()` is designed specifically for books that use one chapter per
```

Comment (hadley) [reply R/bs4_book.R:29]: Oops 😆

**Comment by hadley** [comment]
Can see it in action at https://mastering-shiny.org

File: inst/resources/bs4_book/bs4_book.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    // Insert copy buttons
+    var copyButton = "<div class='copy'><button type='button' class='btn btn-outline-primary btn-copy' title='Copy to clipboard' aria-label='Copy' data-toggle='popover' data-placement='top' data-trigger='hover'>Copy</button></div>";
+    $(copyButton).prependTo("pre");
+    // Initialize tooltips:
+    $('.btn-copy').tooltip({container: 'body', boundary: 'window'});
+
+    // Initialize clipboard:
+    var clipboard = new ClipboardJS('.btn-copy', {
+      text: function(trigger) {
+        return trigger.parentNode.parentNode.textContent;
```
Comment (cderv) [inline inst/resources/bs4_book/bs4_book.js:119]: I think you need this here to solve https://twitter.com/TazPoltorak/status/1321677145404542977?s=20
```suggestion
        return trigger.parentNode.nextSibling.textContent;
```
The chunk code content is the node after the copy button. If you select the all div, you get the 'Copy' text you put in the copy button. If you use an icon like in current bookdown, that would not be an issue..

Comment (hadley) [reply inst/resources/bs4_book/bs4_book.js]: Thanks! I was just about to look into this.

Comment (cderv) [reply inst/resources/bs4_book/bs4_book.js]: I was curious why it worked in `gitbook()` but not in `bs4_book()` so I had a quick look in the morning 😉

**Comment by hadley** [comment]
@yihui this isn't 100% complete, but it's at a point where I think it would be useful to merge into bookdown. Can you please confirm notification of this review request and let me know how long you think it will get back to me?

**Comment by cderv** [comment]
@yihui FYI I already looked at the PR and made some feedbacks to @hadley directly that he included. 

This is pretty independent from the rest of bookdown, with anything new in Suggest. it looks good live: https://mastering-shiny.org/ 

FWIW it looks good to me to merge this first version. IMO we have time to improve as it is used.

**Comment by yihui** [comment]
@hadley I only briefly look at this PR, and I guess I don't need to review the main R code. At the moment, my only question is when **bootstraplib** and **downlit** will be on CRAN, so we could remove `Remotes` in DESCRIPTION, otherwise they will hold the CRAN release of **bookdown**, although there's no hurry at the moment.

**Comment by hadley** [comment]
Thanks @yihui! downlit can go anytime, since it will be accepted by the CRANbot. boostraplib (soon to be bslib) is planned for CRAN release in about a month.

File: R/bs4_book.R [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+  }
+
+  invisible()
+}
+
+tweak_navbar <- function(html, toc, active = "", rmd_index = NULL, repo = NULL) {
+
+  # Source links ------------------------------------------------------------
+  if (!is.null(repo) && active %in% names(rmd_index)) {
+    repo_edit <- paste0(repo, "/edit/master/", rmd_index[[active]])
```
Comment (maelle) [inline R/bs4_book.R:374]: could the default branch not be hardcoded to master?

File: inst/templates/bs4_book.html [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  </div>
+
+</div>
+</div> <!-- .container -->
+
+<footer class="bg-primary text-light mt-5">
+  <div class="container"><div class="row">
+
+  <div class="col-12 col-md-6 mt-3">
+    <p>"<strong>$title$</strong>$if(subtitle)$: $subtitle$$endif$" was written by $for(author)$$author$$sep$, $endfor$. $if(date)$It waslast built on $date$.$endif$</p>
```
Comment (maelle) [inline inst/templates/bs4_book.html:98]:
```suggestion
    <p>"<strong>$title$</strong>$if(subtitle)$: $subtitle$$endif$" was written by $for(author)$$author$$sep$, $endfor$. $if(date)$It was last built on $date$.$endif$</p>
```

**Comment by maelle** [comment]
The new theme fails when a title contains "&" e.g. https://github.com/maelle/mybook

```r
Error in read_xml.raw(charToRaw(enc2utf8(x)), "UTF-8", ..., as_html = as_html,  : 
  xmlParseEntityRef: no name [68]
```

**Comment by maelle** [comment]
Maybe I shouldn't use "&" in titles :thinking: but with https://github.com/rstudio/bookdown/pull/1005 I could

**Comment by hadley** [comment]
Merged; will turn remaining issues on my paper to do list into GitHub issues later today.

File: man/bs4_book.Rd [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+\item{repo}{Link to repository where book is hosted, used to generate
+view source and edit buttons. Currently assumes GitHub and that the book
+is in the root directory of the repo.}
+
+\item{lib_dir, pandoc_args, extra_dependencies, ...}{Passed on to
+[rmarkdown::html_document()].}
+}
+\description{
+This output format is built with [bootstrap](https://getbootstrap.com),
```
[the remark this answers is not in the document]
Comment (cderv) [reply man/bs4_book.Rd:35]: I enabled markdown syntax for roxygen so that this help page renders correctly. Thank you.

**Comment by yihui** [comment]
> boostraplib (soon to be bslib) is planned for CRAN release in about a month.

@hadley Sorry I was not clear last time: I was hoping to merge this PR after **bslib** is on CRAN because its release is unpredictable at the moment. On one hand, I had been told two months ago that its release would be in about a month: https://github.com/rstudio/DT/pull/852#issuecomment-688923812 On the other hand, we are not sure if how many struggles we might have with CRAN for its initial release (you have to be extremely lucky to release a new package in one go).

Anyway, in case I need to make a CRAN release of **bookdown** before **bslib** is on CRAN, I'll cherry-pick the commits.

**Comment by hadley** [comment]
If for some reason you need to do a release before bslib is on CRAN, please talk to me first. I don't think it will be necessary to cherry-pick the commits. I'm also not too worried about initial releases — we have a good process on the tidyverse team and the majority of our recent releases have been accepted on first try.

**Comment by yihui** [comment]
> I don't think it will be necessary to cherry-pick the commits.

So how could I possibly make a release without leaving out the commits related to this PR? I'm just asking, and don't mean I need to make the release (currently there's a minor issue in **blogdown** that needs the next version of **bookdown**, but I can wait).

**Comment by hadley** [comment]
There are at least three options: 

1. Leave out the whole file since all the changes have been scoped to one place.
2. Use a technique to avoid R CMD check detecting bootstraplib usage.
3. Get the shiny team to make a drat repo or similar and use the `Additional_repositories` field

**Pull request merged**